### ISSUE TITLE
monrel - Slashing penalty is unfairly paid by a subset of users if a deficit is accumulated.

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-02-rio-network-core-protocol-judging#363]
monrel

high

# Slashing penalty is unfairly paid by a subset of users if a deficit is accumulated.

## Summary

If a deficit is accumulated in the EigenPodManager due to slashing when ETH is being withdrawn the slashing payment will be taken from the first cohort to complete a withdrawal.

## Vulnerability Detail

A deficit can happen in `podOwnerShares[podOwner]` in the EigenPodManager in the EigenLayer protocol. This can happen if validators are slashed when ETH is queued for withdrawal.

The issue is that this deficit will be paid for by the next cohort to complete a withdrawal by calling `settleEpochFromEigenLayer()`.

In the following code we can see how `epochWithdrawals.assetsReceived` is calculated based on the amount received from the `delegationManager.completeQueuedWithdrawal` call

https://github.com/sherlock-audit/2024-02-rio-network-core-protocol/blob/4f01e065c1ed346875cf5b05d2b43e0bcdb4c849/rio-sherlock-audit/contracts/restaking/RioLRTWithdrawalQueue.sol#L242-L268

```solidity
        uint256 balanceBefore = asset.getSelfBalance();

        address[] memory assets = asset.toArray();
        bytes32[] memory roots = new bytes32[](queuedWithdrawalCount);

        IDelegationManager.Withdrawal memory queuedWithdrawal;
        for (uint256 i; i < queuedWithdrawalCount; ++i) {
            queuedWithdrawal = queuedWithdrawals[i];

            roots[i] = _computeWithdrawalRoot(queuedWithdrawal);
            delegationManager.completeQueuedWithdrawal(queuedWithdrawal, assets, middlewareTimesIndexes[i], true);

            // Decrease the amount of ETH queued for withdrawal. We do not need to validate the staker as
            // the aggregate root will be validated below.
            if (asset == ETH_ADDRESS) {
                IRioLRTOperatorDelegator(queuedWithdrawal.staker).decreaseETHQueuedForUserSettlement(
                    queuedWithdrawal.shares[0]
                );
            }
        }
        if (epochWithdrawals.aggregateRoot != keccak256(abi.encode(roots))) {
            revert INVALID_AGGREGATE_WITHDRAWAL_ROOT();
        }
        epochWithdrawals.shareValueOfAssetsReceived = SafeCast.toUint120(epochWithdrawals.sharesOwed);

        uint256 assetsReceived = asset.getSelfBalance() - balanceBefore;
        epochWithdrawals.assetsReceived += SafeCast.toUint120(assetsReceived);
```

the amount received could be 0 if the deficit is larger than the amount queued for this cohort. See following code in `withdrawSharesAsTokens()` EigenPodManager

https://github.com/Layr-Labs/eigenlayer-contracts/blob/e12b03f20f7dceded8de9c6901ab05cfe61a2113/src/contracts/pods/EigenPodManager.sol#L216C1-L220C14

```solidity
            } else {
                podOwnerShares[podOwner] += int256(shares);
                emit PodSharesUpdated(podOwner, int256(shares));
                return;
            }
```

These users will pay for all slashing penalties instead of it being spread out among all LRT holders.

## Impact

If a deficit is accumulated the first cohort to settle will pay for the entire amount. If they can not cover it fully, they will receive 0 and the following cohort will pay for the rest.

## Code Snippet

https://github.com/sherlock-audit/2024-02-rio-network-core-protocol/blob/4f01e065c1ed346875cf5b05d2b43e0bcdb4c849/rio-sherlock-audit/contracts/restaking/RioLRTWithdrawalQueue.sol#L242-L268

## Tool used

Manual Review
## Recommendation

A potential solution to deal with this is to check if a deficit exists in `settleEpochFromEigenLayer()`. If it exists functionality has to be added that spreads the cost of the penalty fairly among all LRT holders.

**Comment by KupiaSecAdmin** [comment]
Escalate

This has to be considered as Invalid/Low basically because EigenLayer's Slashing contracts do not have any features and all paused, it also won't have any features in its upgrades as well, which means no slashing exists.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> This has to be considered as Invalid/Low basically because EigenLayer's Slashing contracts do not have any features and all paused, it also won't have any features in its upgrades as well, which means no slashing exists.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by 0xmonrel** [comment]
You are misunderstanding the issue. This is about consensus layer slashing.

**Comment by nevillehuang** [comment]
@KupiaSecAdmin Could you link resources/proof to your argument so I can review?

**Comment by KupiaSecAdmin** [comment]
@nevillehuang - https://hackmd.io/@-HV50kYcRqOjl_7du8m1AA/BkOyFwc2T#Out-of-Scope
Here's the docs about upgrades of EigenLayer, which Rio is targeting, hope it helps.

@0xmonrel - If you were meant to say about Ethereum's PoS slashing, as I described [here](https://github.com/sherlock-audit/2024-02-rio-network-core-protocol-judging/issues/190), it should be considered as Low.

**Comment by 0xmonrel** [comment]
The arguments does not apply here. The effect on TVL are irrelevant what matters in this issue is that only 1 cohort of users will pay for the slashing if a deficit happens.

**Comment by nevillehuang** [comment]
@solimander Based on targeted eigen layer contracts it seems it is correct that slashing is currently not applicable to Rio. It is also not stated in the contest details that this will be integrated to rio, so I believe this is invalid based on the following sherlock guideline:

> Future issues: Issues that result out of a future integration/implementation that was not mentioned in the docs/README or because of a future change in the code (as a fix to another issue) are not valid issues.

**Comment by 0xmonrel** [comment]
POS slashing is already live and integrated. Slashing from re-staking is not so it is not out of scope.

**Comment by nevillehuang** [comment]
@0xmonrel Could you point me to the correct resource. This will affect validity of #190 as well

**Comment by 0xmonrel** [comment]
Slashing  is implemented natively through the balance update process. When a validator is slashed the balance is decreased, this is then pushed to EigenLayer through `verifyBalanceUpdates()`.  

Here is the logic where slashing is accounted for and a deficit can happen:

https://github.com/Layr-Labs/eigenlayer-contracts/blob/e12b03f20f7dceded8de9c6901ab05cfe61a2113/src/contracts/pods/EigenPodManager.sol#L207-L219

Look here at the `verifyBalanceUpdates()`

https://github.com/Layr-Labs/eigenlayer-contracts/blob/b6a3a91e1c0c126981de409b00b5fba178503447/src/contracts/pods/EigenPod.sol#L175-L221

Read the comment for a description of what happens when a balance is decreased. 

@solimander can most likely verify that this is correct as he has deep understanding of EigenLayer.

**Comment by solimander** [comment]
That's right, you can find more info on the deficit edge case in the [EigenLayer discord](https://discord.com/channels/1089434273720832071/1187171025478230168/1210639477601603660):

**Comment by nevillehuang** [comment]
Thanks alot guys, I believe this issue is correctly judged given the constraint of a deficit edge case.

**Comment by Czar102** [comment]
I'm planning to reject the escalation and leave the issue as is, unless @KupiaSecAdmin or anyone else has a valid argument not to.

**Comment by Czar102** [comment]
Result:
Medium
Unique

**Comment by sherlock-admin3** [comment]
Escalations have been resolved successfully!

Escalation status:
- [KupiaSecAdmin](https://github.com/sherlock-audit/2024-02-rio-network-core-protocol-judging/issues/363/#issuecomment-2023014162): rejected

**Comment by solimander** [comment]
This is definitely an edge case of an edge case. First, slashing needs to occur while the withdrawal is queued. Second, the operator delegator's pod owner shares needs to be low enough that a deficit is possible. For this reason, considering leaving as-is for now and handling this out-of-protocol.
